lib/ecto/adapters/postgres/connection.ex

```diff
@@ -198,22 +198,25 @@ if Code.ensure_loaded?(Postgrex) do
        end)]
     end
 
-    defp insert_all(rows, counter) do
+    defp insert_all(rows, counter, placeholder_index_map \\ %{}) do
       intersperse_reduce(rows, ?,, counter, fn row, counter ->
-        {row, counter} = insert_each(row, counter)
+        {row, counter} = insert_each(row, counter, placeholder_index_map)
         {[?(, row, ?)], counter}
       end)
       |> elem(0)
     end
 
-    defp insert_each(values, counter) do
+    defp insert_each(values, counter, placeholder_index_map \\ %{}) do
       intersperse_reduce(values, ?,, counter, fn
         nil, counter ->
           {"DEFAULT", counter}
 
         {%Ecto.Query{} = query, params_counter}, counter ->
           {[?(, all(query), ?)], counter + params_counter}
 
+        {:placeholder, placeholder_key}, counter ->
```

Comment:
On unzip_inserts, make it so this is passed as `{:placeholder, placeholder_index}`, this way you don't need to pass the `placeholder_index_map` to this function.